### PR TITLE
Move prepare code to its own package

### LINKED ISSUE #450: Move prepare code to standalone package
#### Proposal

The Cluster preparation code should not live in the k8s client.

It should have its own package.

#### Background

Implementation code shouldn't rely on this client.

### PR DESCRIPTION
**Pull request by dtomcej**
This PR:

- Moves the prepare code to its own package
- Implements more interfaces for the k8s Client
- Updates the integration tests to use the k8s client interface

Fixes #450 

Note: The `k8s.NewClient` now returns an interface, so you cannot interact with the underlying wrapper anymore, so this enforces the interface use.

### CODE REVIEW

File: pkg/prepare/prepare.go [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+// Ensure the client wrapper fits the Client interface
+var _ Interface = (*Prepare)(nil)
+
+// Prepare holds the clients for the various resource controllers.
+type Prepare struct {
+	client k8s.Client
+}
+
+// NewPrepare returns an initialized prepare object.
+func NewPrepare(client k8s.Client) Interface {
```
[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/prepare/prepare.go]: Returning an interface ensures that the implementation isn't referenced.

File: pkg/prepare/prepare.go [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	supportedCoreDNSVersions = []string{
```
[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/prepare/prepare.go]: `syntax error: non-declaration statement outside function body`

It doesn't like the slice. Can't be a constant.

File: pkg/k8s/client.go [hunk 3]
```diff
@@ -589,54 +135,9 @@ func buildSmiSplitClient(config *rest.Config) (*splitClient.Clientset, error) {
...
-// UpdateDeployment updates the specified deployment.
-func (w *ClientWrapper) UpdateDeployment(deployment *appsv1.Deployment) (*appsv1.Deployment, error) {
-	return w.kubeClient.AppsV1().Deployments(deployment.Namespace).Update(deployment)
-}
-
-// translateNotFoundError will translate a "not found" error to a boolean return
+// TranslateNotFoundError will translate a "not found" error to a boolean return
 // value which indicates if the resource exists and a nil error.
-func translateNotFoundError(err error) (bool, error) {
+func TranslateNotFoundError(err error) (bool, error) {
```
[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/k8s/client.go]: I think we should discuss this one in a separate PR, since it is used a bit.

[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/k8s/client.go]: https://github.com/containous/maesh/issues/457

File: pkg/prepare/prepare.go [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+		"1.3",
+		"1.4",
+		"1.5",
+		"1.6",
+	}
+)
+
+// Interface is an interface for the prepare methods.
+type Interface interface {
```
[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/prepare/prepare.go]: We test them, and use the 3 methods separately.

StartInformers is still used, as we check and start informers in the main controller command.

File: pkg/prepare/prepare.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jspdown) [reply pkg/prepare/prepare.go]: It does seems to work: https://goplay.space/#sFrVNxomfY3

[the remark this answers is not in the document]
Comment (dtomcej) [reply pkg/prepare/prepare.go]: ahhh I see, I removed the var too

**Pull request merged**